solr/core/src/java/org/apache/solr/core/SolrCores.java

```diff
@@ -99,37 +85,34 @@ protected void close() {
     waitForLoadingCoresToFinish(30 * 1000);
     Collection<SolrCore> coreList = new ArrayList<>();
 
-    // Release transient core cache.
-    synchronized (modifyLock) {
-      if (transientSolrCoreCacheFactory != null) {
-        getTransientCacheHandler().close();
-      }
-    }
-
     // It might be possible for one of the cores to move from one list to another while we're
     // closing them. So loop through the lists until they're all empty. In particular, the core
     // could have moved from the transient list to the pendingCloses list.
-    do {
-      coreList.clear();
+    while (true) {
+
       synchronized (modifyLock) {
-        // make a copy of the cores then clear the map so the core isn't handed out to a request
-        // again
-        coreList.addAll(cores.values());
-        cores.clear();
-        if (transientSolrCoreCacheFactory != null) {
-          coreList.addAll(getTransientCacheHandler().prepareForShutdown());
+        // remove all loaded cores; add to our working list.
+        for (String name : getLoadedCoreNames()) {
+          final var core = remove(name);
+          if (core != null) {
```

Comment:
not sure I follow the null check here. both 'getLoadedCoreNames' and 'remove' happen under lock, what is the scenario where this would be null? also, please reconsider the use of dedicated methods for basically 'keySet' and 'remove' with extra locking, it is adding a bunch of complexity to this code that is not needed.

Reply:
I suppose core can't be null; I just added the check to be sure.  I'll change it to an assertion.  
What complexity is in SolrCores.close() that isn't needed?  I suspect maybe you mean the total work that is done (i.e. adding code that these methods call), but I don't think that should _count_ in terms of complexity of this method -- being able to understand what it does.

If your point wasn't actually complexity but was about performance, I think it's a trivial matter.  Consider a SolrCores is closed once :-) and I disagree it'd be noticeably slower in any benchmark.

Reply:
I meant cognitive complexity. I needed to bounce around a bit to verify what those methods do only to find that they are calling the Map methods (and so realized that the null check was not needed). what is the need for extra methods when they don't do anything else, makes the code more difficult to understand.

Reply:
Firstly, I don't think SolrCore.close calling getLoadedCoreNames adds much cognitive complexity since its name is clear and it does what you'd think it does.  I think you are asking, why does `getLoadedCoreNames` exist at all if it's so trivial.  It existed before but the reason is to allow a subclass that could implement the concept of transient cores, which is the point of this PR.

Reply:
I agree with your reasoning, thanks for the clarifications!

Reply:
The null check is actually needed, as I found TestCoreDiscovery.testTooManyTransientCores will fail.  It's vaguely explained in a comment here.  In more detail, transient core cache eviction can happen on any core manipulation, including the removal call.  So we might get core list A, B, C.  Then we remove A, triggering a cache maintenance resulting in B getting removed/evicted.  Then we try to remove B and it's already removed (and was put in pendingCloses).